src/v/cloud_roles/azure_vm_refresh_impl.cc

```diff
@@ -0,0 +1,126 @@
+  net::unresolved_address address,
+  aws_region_name region,
+  ss::abort_source& as,
+  retry_params retry_params)
+  : refresh_credentials::impl(
+    std::move(address), std::move(region), as, retry_params) {}
+
+auto azure_vm_refresh_impl::print(std::ostream& os) const -> std::ostream& {
+    fmt::print(os, "azure_vm_refresh_impl{{address:{}}}", address());
+    return os;
+}
+
+auto azure_vm_refresh_impl::fetch_credentials() -> ss::future<api_response> {
+    auto client_id_opt = config::shard_local_cfg()
+                           .cloud_storage_azure_managed_identity_id.value();
+    if (unlikely(!client_id_opt.has_value())) {
+        // This implementation requires client_id from config to be set.
+        // Strictly speaking, IMDS service does not require it if there is only
+        // a system-assigned managed identity or only one user-assigned managed
+        // identity but it would be brittle to not specify it, it could break if
+        // the user added a new one. This verification is also performed in
+        // admin/server.cc::patch_cluster_config
+        vlog(
+          clrl_log.error,
+          "missing cloud_storage_azure_managed_identity_id in config");
+        // return value is not a perfect match but it works
+        co_return api_request_error{
+          .status = boost::beast::http::status::bad_request,
+          .reason = "missing cloud_storage_azure_managed_identity_id",
+          .error_kind = api_request_error_kind::failed_abort};
+    }
+
+    // performs this GET
+    // 'http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https://storage.azure.com/&client_id={$client_id}'
+    // to the vm-local IMDS
+    auto req = http::client::request_header{};
+    req.method(boost::beast::http::verb::get);
+    // TODO fix deps and use boost::url
+    req.target(fmt::format(
+      "/metadata/identity/oauth2/token?"
+      "api-version=2018-02-01"
+      "&resource=https%3A%2F%2Fstorage.azure.com%2F"
+      "&client_id={}",
+      client_id_opt.value()));
+    req.set("Metadata", "true");
+
+    co_return co_await make_request(
+      co_await make_api_client("azure_vm_instance_metadata"), std::move(req));
+}
+
+auto azure_vm_refresh_impl::parse_response(iobuf response)
+  -> api_response_parse_result {
+    // simplified schema for (an example)
+    //  {
+    //    "token_type": "Bearer",
+    //    "expires_in": "3599",
+    //    "access_token":
+    //    "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6Ik1uQ19WWmNBVGZNNXBP..."
+    //  }
+    // note that the expires_in field is a string, not an integer, and that the
+    // complete response has more fields but we just care for these 3
+    constexpr static auto success_schema_str = R"json(
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "type": "object",
+  "properties": {
+    "token_type": {
+      "type": "string",
+      "pattern": "Bearer"
+    },
+    "expires_in": {
+      "type": "string"
```

Comment:
suggestion: this pattern could be a regex specifying a sequence of digits, as we know it is an int serialized as a string, this would make the later case to int64_t safer

Reply:
good idea, something like 
^\+?\d{1,18}$

to match zero or one '+' and 1 up to 18 digits after that